R/prep_data_utils.R

```diff
@@ -64,3 +64,79 @@ find_new_missingness <- function(d, recipe) {
   new_missing <- dplyr::setdiff(missing_now, missing_then)
   return(dplyr::intersect(new_missing, predictors))
 }
+
+#' @title Convert character date columns to dates
+#'
+#' @description I try to convert columns ending in "DTS" to type Date. I'll take
+#' my best guess at the format and return a more standard one if I can. Times
+#' will be removed.
+#'
+#' @param d A dataframe or tibble containing data to try to convert to dates.
+#'
+#' @return A tibble containing the converted date columns. If no columns
+#' needed conversion, the original data will be returned.
+#' @import purrr
+#' @importFrom lubridate guess_formats date
+#' @export
+#'
+#' @examples
+#' d <- tibble::tibble(a_DTS = c("2018-3-25", "2018-3-25"),
+#' b_nums = c(2, 4),
+#' c_DTS = c("03-25-2018", "03-25-2018"),
+#' d_chars = c("a", "b"),
+#' e_date = lubridate::mdy(c("3-25-2018", "3-25-2018")))
+#'
+#' d <- convert_date_cols(d)
+#'
+convert_date_cols <- function(d) {
+
+  # Extract character date columns only
+  dd <- d[1, find_date_cols(d)]
+  col_names <- names(dd[!(map_lgl(dd, is.Date))])
+  dd <- dd[1, col_names]
+
+  date_formats <- map(dd,
+                      guess_formats,
+                      orders = c("ymd", "mdy", "ymd HMS", "mdy HMS"))
+
+  # Find working formats
+  valid_formats <- map2(dd, date_formats, function(x, y) {
```

Comment:
When I get here from the input below, dd and date_formats have different length, which presumably means the shorter one gets recycled, but I don't think we want to do that (and I'm surprised purrr allows it). Also, surely we don't need nested `map2` loops, right? That's an awfully complicated structure. Also, use `map_lgl` rather than `unlist(map))`, it's much safer.

```
td <- data.frame(aDTS = c("2017-01-02", "2018-10-01"),
                     actual_date = as.Date(c("2018-01-02", "2018-03-04")),
                 chr_var = c("a", "b"))
prep_data(td, make_dummies = FALSE)
```